### PR TITLE
Muriel dev

### PR DESCRIPTION
**Pull request by ThatIsAUsername**
Quality of life improvements so Muriel doesn't get in her way as much.

### CODE REVIEW

File: src/AI/AIUtils.java [hunk 1]
```diff
@@ -70,6 +84,18 @@
...
   public static Map<UnitActionFactory, ArrayList<GameAction> > getAvailableUnitActionsByType(Unit unit, GameMap gameMap)
+  {
+    boolean includeOccupiedDestinations = false;
+    return getAvailableUnitActionsByType(unit, gameMap, includeOccupiedDestinations);
+  }
+  /**
+   * Finds all actions available to unit, and organizes them by type instead of by location.
+   * Assumes caller isn't interested in moving into units' current spaces
+   * @param unit The unit under consideration.
+   * @param gameMap The world in which the Unit lives.
```
Comment (ThislsAUsername) [inline src/AI/AIUtils.java:95]: *cough*

File: src/Engine/GameAction.java [hunk 2]
```diff
@@ -30,6 +30,7 @@
    * Returns a GameEventQueue with the events that make up this action. If the action
    * was constructed incorrectly, this should return an empty GameEventQueue.
    */
+  public abstract Unit getActor();
```
Comment (ThislsAUsername) [inline src/Engine/GameAction.java:33]: Why didn't you just add a default implementation?

File: src/AI/Muriel.java [hunk 3]
```diff
@@ -219,269 +239,441 @@ public GameAction getNextAction(GameMap gameMap)
...
+    // Flag the actor as an obstacle so the unit it's pushing can't push back.
+    if( !unitSelector.isObstacle(actor) )
+      unitSelector.flagObstacle(actor);
+    obstructedActions.put(actor, desiredAction, actionValue);
+
+    // Give the obstacle priority so it has to move first.
+    unitSelector.flagObstacle(obstacle);
+  }
+  /** Compare ga against any currently-obstructed actions. Return true if it is better
+   * than any of those, and false if an existing known action should take priority. */
```
Comment (ThislsAUsername) [inline src/AI/Muriel.java:315]:
```suggestion
  /** Compare input action against known blocked actions doable from that space. Return true if it is better
   * than any of those, and false if an existing known action should take priority. */
```

File: src/Test/TestAIMuriel.java [hunk 4]
```diff
@@ -168,4 +172,134 @@ private boolean testClearAttackRoute()
...
   }
 
+  /** Two units want to go the same way. One is where the other should end up. Make sure they coordinate. */
+  /** START: [ ] [ ] INF [ ] [ ] INF [ ] [ ] [ ] HQ  */
+  /** GOAL : [ ] [ ] [ ] [ ] [ ] INF [ ] [ ] INF HQ */
+  private boolean testWalkInLine()
+  {
+    // Generate a custom tiny map to keep things simple.
+    TerrainType[] HQCol = {TerrainType.HEADQUARTERS};
+    TerrainType[] GRCol = {TerrainType.GRASS};
```
Comment (ThislsAUsername) [inline src/Test/TestAIMuriel.java:182]: Pretty sure we have 2-character abbreviations already defined...

File: src/Test/TestAIMuriel.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+    testPassed &= validate(lt.x == 5, "    Muriel failed to move left infantry as intended!");
+    testPassed &= validate(rt.x == 8, "    Muriel failed to move right infantry as intended!");
+
+    // Clean up
+    cleanupTest();
+    return testPassed;
+  }
+
+  /** Two units want to go the same way. One is where the other should end up. Make sure they coordinate. */
```
Comment (ThislsAUsername) [inline src/Test/TestAIMuriel.java:215]:
```suggestion
  /** This inf needs to wade through a sea of Md Tanks to get to the HQ. */
```

**Comment by ThislsAUsername** [comment]
Looks pretty good. If you're about done, it might be simpler to get this merged before I get too deep into fixing Wally.

File: src/Engine/GameAction.java (hunk 2 above)
Comment (ThatIsAUsername) [reply src/Engine/GameAction.java:33]: I dum

File: src/Test/TestAIMuriel.java (hunk 4 above)
Comment (ThatIsAUsername) [reply src/Test/TestAIMuriel.java:182]: We do, but for some reason they are protected members of `IMapBuilder`, so I just did this instead. We could just make those public unless you see a reason not to; they are `static final` anyway, so I'm not really sure why we wouldn't.

Comment (ThislsAUsername) [reply src/Test/TestAIMuriel.java:182]: I don't see any reason to make them non-public either, lol

File: src/Engine/GameAction.java (hunk 2 above)
Comment (ThislsAUsername) [reply src/Engine/GameAction.java]: Not necessarily. Having to provide an implementation means you have to think about what it means for each subclass, which might help avoid surprises down the line.
I'm generally in favor of terseness, but letting the compiler help you keep track of stuff can also help.

**Comment by ThatIsAUsername** [comment]
> If you're about done, it might be simpler to get this merged before I get too deep into fixing Wally.

I don't think I'm making any more significant changes at this point. At some point I may move the AI tests into a separate file, run them on all AIs, and set per-AI thresholds on what constitutes "passing" to prevent regressions, and I definitely have more improvements to eventually make, but that's no hurry.

**Pull request merged**